transports/webrtc/src/transport.rs

```diff
@@ -0,0 +1,697 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use libp2p_core::{
+    multiaddr::{Multiaddr, Protocol},
+    transport::{ListenerEvent, TransportError},
+    Transport,
+};
+
+use futures::{
+    channel::{mpsc, oneshot},
+    future::BoxFuture,
+    prelude::*,
+    ready, TryFutureExt,
+};
+use futures_timer::Delay;
+use if_watch::{IfEvent, IfWatcher};
+use log::{debug, error, trace};
+use tinytemplate::TinyTemplate;
+use tokio_crate::net::{ToSocketAddrs, UdpSocket};
+use webrtc::api::setting_engine::SettingEngine;
+use webrtc::api::APIBuilder;
+use webrtc::data_channel::data_channel_init::RTCDataChannelInit;
+use webrtc::peer_connection::certificate::RTCCertificate;
+use webrtc::peer_connection::configuration::RTCConfiguration;
+use webrtc::peer_connection::sdp::session_description::RTCSessionDescription;
+use webrtc_data::data_channel::DataChannel as DetachedDataChannel;
+use webrtc_ice::network_type::NetworkType;
+use webrtc_ice::udp_mux::UDPMux;
+use webrtc_ice::udp_network::UDPNetwork;
+
+use std::borrow::Cow;
+use std::io;
+use std::net::IpAddr;
+use std::net::SocketAddr;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex};
+use std::task::{Context, Poll};
+use std::time::Duration;
+
+use crate::error::Error;
+use crate::sdp;
+
+use crate::connection::Connection;
+use crate::udp_mux::UDPMuxNewAddr;
+use crate::udp_mux::UDPMuxParams;
+use crate::upgrade::WebRTCUpgrade;
+
+enum IfWatch {
+    Pending(BoxFuture<'static, io::Result<IfWatcher>>),
+    Ready(IfWatcher),
+}
+
+/// The listening addresses of a [`WebRTCTransport`].
+enum InAddr {
+    /// The stream accepts connections on a single interface.
+    One { out: Option<Multiaddr> },
+    /// The stream accepts connections on all interfaces.
+    Any { if_watch: IfWatch },
+}
+
+/// A WebRTC transport with direct p2p communication (without a STUN server).
+#[derive(Clone)]
+pub struct WebRTCTransport {
+    /// A `RTCConfiguration` which holds this peer's certificate(s).
+    config: RTCConfiguration,
+
+    /// The `UDPMux` that manages all ICE connections.
+    udp_mux: Arc<dyn UDPMux + Send + Sync>,
+
+    /// The local address of `udp_mux`.
+    udp_mux_addr: SocketAddr,
+
+    /// The receiver for new `SocketAddr` connecting to this peer.
+    new_addr_rx: Arc<Mutex<mpsc::Receiver<Multiaddr>>>,
+}
+
+impl WebRTCTransport {
+    /// Create a new WebRTC transport.
+    ///
+    /// Creates a UDP socket bound to `listen_addr`.
+    pub async fn new<A: ToSocketAddrs>(
+        certificate: RTCCertificate,
+        listen_addr: A,
+    ) -> Result<Self, TransportError<Error>> {
+        // bind to `listen_addr` and construct a UDP mux.
+        let socket = UdpSocket::bind(listen_addr)
+            .map_err(Error::IoError)
+            .map_err(TransportError::Other)
+            .await?;
```

Comment:
It seems a bit weird that we start the listening process in the constructor already when we also have the `listen_on` abstraction.

I've had a similar problem when I was trying to fit a Tor-based transport into `rust-libp2p`. I assume the problem is that you need a specific certificate per listen-address?

One thing that we could do is pass a map of certificates in the constructor and check the map for an entry within `listen_on`. If the correct certificate for this address is not present, we can fail during `listen_on`. If it is present, we can continue with the certificate.

Reply:
> I assume the problem is that you need a specific certificate per listen-address?

That, but there's also the `UDPMux` which is used to dial an address. So both `listen_on` and `do_dial` share the same "socket". In theory, if `listen_on` was `mut self`, I could construct `udp_mux` and use it in `do_dial`, but `listen_on` is not mutable.

> One thing that we could do is pass a map of certificates in the constructor and check the map for an entry within listen_on. If the correct certificate for this address is not present, we can fail during listen_on. If it is present, we can continue with the certificate.

Reply:
Referencing https://github.com/libp2p/rust-libp2p/pull/2652 here which might make things easier.

Reply:
> > I assume the problem is that you need a specific certificate per listen-address?
> 
> That, but there's also the `UDPMux` which is used to dial an address. So both `listen_on` and `do_dial` share the same "socket".

That is definitely something that doesn't fit particularly well into our abstractions.

> In theory, if `listen_on` was `mut self`, I could construct `udp_mux` and use it in `do_dial`, but `listen_on` is not mutable.

On latest master, it already takes `&mut self`!